### PR TITLE
[typescript-fetch] Support deepObject serialization in query parameters

### LINKED ISSUE #2233: [BUG][typescript-fetch] deepObject serialization not supported in query string
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

A query parameter of `type: object` using `additionalProperties` produces Typescript code that fails to compile.

A similar parameter without `additionalProperties` that instead specifies properties compiles properly, but serializes the parameter improperly.

###### Typescript compiler error with `additionalProperties`

```
api/test/apis/DefaultApi.ts
Type error: Type '{ [key: string]: number; }' is not assignable to type 'string | number | boolean | (string | number | boolean)[]'.
  Type '{ [key: string]: number; }' is not assignable to type 'string'.  TS2322

    30 | 
    31 |         if (requestParameters.foffset !== undefined) {
  > 32 |             queryParameters['foffset'] = requestParameters.foffset;
       |             ^
    33 |         }
    34 | 
    35 |         const headerParameters: runtime.HTTPHeaders = {};
```

###### Invalid serialization with specified properties

The parameter is decoded to the string representation of an object in Javascript:

`foffset=%5Bobject%20Object%5D` (URL encoded)  
`foffset=[object Object]` (URL decoded)

##### openapi-generator version

v4.0.0-beta2

As far as I can tell, there have been no relevant changes on `master`.

##### OpenAPI declaration file

```yaml
openapi: 3.0.0
info:
  title: Test API
  version: 0.0.1
paths:
  /search:
    get:
      parameters:
        - name: foffset
          in: query
          required: false
          schema:
            type: object
            additionalProperties:
              type: integer
      responses:
        '200':
          description: OK

```

##### Command line used for generation

`openapi-generator generate -g typescript-fetch -i ./test.openapi.yml -o api/` 

##### Steps to reproduce

1. Specify a query parameter of `type: object`.
2. Generate `typescript-fetch` client code.
3. Include/compile the generated API typescript file (`additionalProperties` variant fails here).
4. Perform an API request providing an object for the parameter.

##### Related issues/PRs

Possibly related to PR #569 and Issue #802.

##### Suggest a fix

I'm working on a first pull request.  
The idea is to allow nested `HTTPQuery` objects that are recursively serialized.

### PR DESCRIPTION
**Pull request by Dijky**
Fixes #2233.

@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @nicokoenig @topce 

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh`, `./bin/security/{LANG}-petstore.sh` and `./bin/openapi3/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Add support for nested objects in query parameters.

Object type query parameters can currently not be properly serialized (see related issue above).  
This PR allows nested object parameters in the `HTTPQuery` type and extends the `querystring` function of the API runtime library to recursively serialize objects with the `deepObject` style as defined in [OpenAPI 3.0](https://swagger.io/docs/specification/serialization/#query).

This method does *not* support arrays of objects, as this would cause an ambiguous serialization due to the use of the `explode` format for arrays.  
E.g. a query string `person[name]=Alice&person[name]=Bob` could be interpreted as an array of person objects with one name property each, or one person object with an array property `name`.

Feedback is highly appreciated.

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache [hunk 1]
```diff
@@ -183,15 +183,19 @@ export function exists(json: any, key: string) {
     return value !== null && value !== undefined;
 }
 
-export function querystring(params: HTTPQuery) {
+export function querystring(params: HTTPQuery, prefix: string = '') {
     return Object.keys(params)
         .map((key) => {
+            const fullKey = prefix + (prefix.length ? `[${key}]` : key);
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache]: maybe it should be configurable whether to use square brackets `[` `]` or e.g. dots `.`?

Comment (Dijky) [reply modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache]: The format I chose matches the `deepObject` style as it is [specified](https://swagger.io/docs/specification/serialization/#query) in OpenAPI 3.0.  
While I have nothing against other formats, that would IMO be a topic to discuss with the OAI.

AFAIK, this generator currently only supports the `form` style with `explode` enabled and no other combination, so there is a general lack of configurability for query parameters.

**Comment by Dijky** [comment]
I just ran the scripts to update the sample files and commited the results.

Some changes in the generated samples don't look like they come from my work, so I'm not quite sure what to do with them. I commited them separately in ae1aab9.

File: modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache [hunk 2]
```diff
@@ -183,15 +183,19 @@ export function exists(json: any, key: string) {
...
             const value = params[key];
             if (value instanceof Array) {
-                const multiValue = value.join(`&${encodeURIComponent(key)}=`);
-                return `${encodeURIComponent(key)}=${multiValue}`;
+                const multiValue = value.join(`&${encodeURIComponent(fullKey)}=`);
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache]:
```suggestion
                const multiValue = value.map(singleValue => encodeURIComponent(singleValue))
                .join(`&${encodeURIComponent(fullKey)}=`);
```

**Comment by Dijky** [comment]
@macjohnny Nice catch! Interestingly, the encoding of array elements is omitted in the current master as well.

I noticed that the weird changes to the sample files are caused by the scripts. The OpenAPI 2.0 and 3.0 scripts generate different samples to the same folders, so the ultimate outcome depends on the order of script execution.  
This has nothing to do with my changes, so I removed the changes and cleaned up the PR.

**Comment by Dijky** [comment]
@macjohnny Thanks für approving my other PR.  
Do you have any further comment on this one here?

Also, please remove the WIP label if you can. It was added by auto-labeler and is not correct any more.  
This PR is ready from my end.

**Comment by macjohnny** [comment]
cc @wing328

**Comment by wing328** [comment]
Sorry for the delay. Missed the notification.

**Pull request merged**